### PR TITLE
Resolving names with underscores

### PR DESCRIPTION
**Pull request by sigurdm**
Also quite some refactoring of name-resolution.

### CODE REVIEW

**Comment by sigurdm** [comment]
cc @nichite

File: lib/names.dart [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import 'package:protobuf/meta.dart';
 import 'package:protoc_plugin/src/dart_options.pb.dart';
 import 'package:protoc_plugin/src/descriptor.pb.dart';
+import 'dart:io';
```
Comment (sigurdm) [inline lib/names.dart]: Remove unused import

File: lib/names.dart [hunk 2]
```diff
@@ -32,52 +33,49 @@ class MemberNames {
...
+    i++;
   }
+}
 
-  String s = _fileNameWithoutExtension(descriptor).replaceAll('-', '_');
+/// Chooses the name of the Dart class holding top-level extensions.
+String extensionClassName(
+    FileDescriptorProto descriptor, Set<String> usedNames) {
+  String s = avoidInitialUnderscore(
+      _fileNameWithoutExtension(descriptor).replaceAll('-', '_'));
```
Comment (sigurdm) [inline lib/names.dart]: Other invalid characters can occur in filenames

File: lib/names.dart [hunk 3]
```diff
@@ -94,24 +92,71 @@ class DartNameOptionException implements Exception {
...
 
+Iterable<String> defaultSuffixGenerator() sync* {
+  int i = 0;
+  while (true) {
+    yield '_$i';
+    i++;
+  }
+}
+
+/// Returns a [name] that is not contained in usedNames by suffixing it with the
```
Comment (sigurdm) [inline lib/names.dart]: highlight usedNames

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,27 @@
+## 0.12.0
```
Comment (szakarias) [inline CHANGELOG.md]: Update the version in the pubspec too.

File: lib/names.dart [hunk 5]
```diff
@@ -94,24 +92,71 @@ class DartNameOptionException implements Exception {
...
+  usedNames.addAll(candidateVariants);
+  return '$name$usedSuffix';
+}
+
+Iterable<String> _messageOrEnumClassSuffixes() sync* {
+  yield '_';
+  int i = 0;
+  while (true) {
+    yield ('_$i');
+  }
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/names.dart]: !!!!
Done

File: lib/names.dart [hunk 6]
```diff
@@ -319,6 +365,23 @@ bool _isDartFieldName(name) => name.startsWith(_dartFieldNameExpr);
 
 final _dartFieldNameExpr = new RegExp(r'^[a-z]\w+$');
 
+/// Names that would collide with capitalized core dart names as top-level
+/// identifiers.
+final List<String> toplevelReservedCapitalizedNames = const <String>[
+  'List',
+  'Function'
+];
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/names.dart]: Done

File: test/leading_underscores_test.dart [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test/test.dart';
+import 'package:fixnum/fixnum.dart';
+
+import '../out/protos/_leading_underscores.pb.dart';
+
+void main() {
+  test('can set, read and all fields and refer to types', () {
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply test/leading_underscores_test.dart]: fixed

File: test/protos/_leading_underscores.proto [hunk 8]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply test/protos/_leading_underscores.proto]: Done

File: lib/names.dart [hunk 9]
```diff
@@ -319,6 +365,23 @@ bool _isDartFieldName(name) => name.startsWith(_dartFieldNameExpr);
 
 final _dartFieldNameExpr = new RegExp(r'^[a-z]\w+$');
 
+/// Names that would collide with capitalized core dart names as top-level
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply lib/names.dart]: Done

File: CHANGELOG.md (hunk 4 above)
Comment (sigurdm) [reply CHANGELOG.md]: Done

**Pull request merged**